facebook/OpenBIC
# fby4: ff: Support get/set state effecter

Description
- Support switch UART between BIC and CXL. Implement PLDM_switch_UART handler to get the request from BMC.
- Support BMC to get/set GPIO of BIC

Motivation:
- Support get/set state effecter.

Test plan:
- Build code: Pass
- Switch UART: Pass
- Get/set GPIO: Pass

Log:
1. Switch UART.
- Switch to CXL console root@bmc:~# pldmtool raw -d 0x80 0x02 0x39 0x3 0x0 0x1 0x01 0x01 -m 42 pldmtool: Tx: 84 02 39 03 00 01 01 01
pldmtool: Rx: 04 02 39 00
root@bmc:~# obmc-console-client -i host3

uart:~$ help
Please press the <Tab> button to see all available commands. You can also use the <Tab> button to prompt or auto-complete all commands or its subcommands. You can try to call commands with <-h> or <--help> parameter for more information.

Shell supports following meta-keys:
  Ctrl + (a key from: abcdefklnpuw)
  Alt  + (a key from: bf)
Please refer to shell documentation for more details.

Available commands:
  bld_ver    :Version info commands
  clear      :Clear screen.
  coredump   :Coredump commands (flash partition backend)
  cxl        :cxl commands
  ddr        :DDR Commands
  device     :Device commands
  devmem     :Read/write physical memory"devmem address [width [value]]
  events     :Vistara Events shell commands
  help       :Prints the help message.
  history    :Command history.
  i2c        :I2C commands
  kernel     :Kernel commands
  membridge  :Membridge Commands
  mfio       :MFIO commands
  pcie_ep    :PCIe EP commands
  pyarctic   :PyArctic Tool Commands
  qspi       :Vistara QSPI shell commands
  reset      :Reset commands
  resize     :Console gets terminal screen size or assumes default in case the
              readout fails. It must be executed after each terminal width
              change to ensure correct text display.
  select     :Selects new root command. In order for the command to be selected,
              it must meet the criteria:
              - it is a static command
              - it is not preceded by a dynamic command
              - it accepts arguments
              Return to the main command tree is done by pressing alt+r.
  sensor     :Sensor commands
  shell      :Useful, not Unix-like shell commands.
  spd        :SPD commands
  vistara    :Vistara shell commands
  vtmon      :AVS VTMON commands
uart:~$

- Switch to BIC console root@bmc:~# pldmtool raw -d 0x80 0x02 0x39 0x3 0x0 0x1 0x01 0x00 -m 42 pldmtool: Tx: 84 02 39 03 00 01 01 00
pldmtool: Rx: 04 02 39 00
root@bmc:~# obmc-console-client -i host3

uart:~$ help
Please press the <Tab> button to see all available commands. You can also use the <Tab> button to prompt or auto-complete all commands or its subcommands. You can try to call commands with <-h> or <--help> parameter for more information.

Shell supports following meta-keys:
  Ctrl + (a key from: abcdefklnpuw)
  Alt  + (a key from: bf)
Please refer to shell documentation for more details.

Available commands:
  adc       :ADC commands
  clear     :Clear screen.
  date      :Date commands
  device    :Device commands
  flash     :Flash shell commands
  gpio      :GPIO commands
  help      :Prints the help message.
  history   :Command history.
  hwinfo    :HWINFO commands
  i2c       :I2C commands
  i3c       :I3C commands
  jtag      :JTAG shell commands
  kernel    :Kernel commands
  log       :Commands for controlling logger
  md        :Mem Display command
  mw        :Mem Write command
  platform  :Platform commands
  resize    :Console gets terminal screen size or assumes default in case the
             readout fails. It must be executed after each terminal width change
             to ensure correct text display.
  shell     :Useful, not Unix-like shell commands.
uart:~$

2. Control GPIO.
- Get GPIO 0x32 *BMC console
root@bmc:~# pldmtool raw -d 0x80 0x02 0x3a 0x23 0xff -m 42 pldmtool: Tx: 84 02 3a 23 ff
pldmtool: Rx: 04 02 3a 00 02 01 02 02 01 01 01

*BIC console
uart:~$ platform gpio get 35
[35 ] USB_MUX1_SEL_R                     : PP  | output(O) | 0(0)

- Set GPIO 0x32 to high root@bmc:~# pldmtool raw -d 0x80 0x02 0x39 0x23 0xff 0x02 0x00 0x00 0x00 0x02 -m 42 pldmtool: Tx: 86 02 39 23 ff 02 00 00 00 02
pldmtool: Rx: 06 02 39 00

- Get GPIO 0x32 *BMC console
root@bmc:~# pldmtool raw -d 0x80 0x02 0x3a 0x23 0xff -m 42 pldmtool: Tx: 87 02 3a 23 ff
pldmtool: Rx: 07 02 3a 00 02 01 02 02 01 02 02

*BIC console
uart:~$ platform gpio get 35
[35 ] USB_MUX1_SEL_R                     : PP  | output(O) | 1(1)
